File: vision/branches/src/java/cast/core/CASTComponentPermissionsMap.java
package cast.core;

import java.util.Hashtable;

import cast.cdl.WorkingMemoryPermissions;


public class CASTComponentPermissionsMap {

	
	private static class PermAndCheck extends Pair<WorkingMemoryPermissions, Boolean> {
		PermAndCheck(WorkingMemoryPermissions _a, boolean _b) {
			super(_a, _b);
		}
	}
	
	private final Hashtable<String, PermAndCheck> m_localPermissions;
	private final Hashtable<String, Hashtable<String, PermAndCheck>> m_globalPermissions;
	
	public CASTComponentPermissionsMap(String _subarch) {
		//keep local permissions separate for ease
		m_localPermissions = new Hashtable<String, PermAndCheck>();
		m_globalPermissions = new Hashtable<String, Hashtable<String,PermAndCheck>>();
		//store local permissions in global map for generality
		m_globalPermissions.put(_subarch, m_localPermissions);
	}	
	
	public WorkingMemoryPermissions setPermissions(String _id, WorkingMemoryPermissions _permissions) {
		PermAndCheck check =  m_localPermissions.put(_id, new PermAndCheck(_permissions,true));
		if(check == null) {
			return null;
		}
		else {
			return check.m_first;
		}
	}
	
	public WorkingMemoryPermissions setPermissions(String _id, String _subarch, WorkingMemoryPermissions _permissions) {
		if(!m_globalPermissions.containsKey(_subarch)) {
			m_globalPermissions.put(_subarch, new Hashtable<String, PermAndCheck>());
		}
		PermAndCheck check = m_globalPermissions.get(_subarch).put(_id, new PermAndCheck(_permissions,true));
		
		if(check == null) {
			return null;
		}
		else {
			return check.m_first;
		}
	}

	public WorkingMemoryPermissions getPermissions(String _id) {
		return m_localPermissions.get(_id).m_first;
	}
	
	public WorkingMemoryPermissions getPermissions(String _id, String _subarch) {
		if(!m_globalPermissions.containsKey(_subarch)) {
			return null;
		}
		PermAndCheck check = m_globalPermissions.get(_subarch).get(_id);
		if(check == null) {
			return null;
		}
		else {
			return check.m_first;
		}
		
	}

	public boolean hasPermissions(String _id) {
		return getPermissions(_id) != null;
	}
	
	public boolean hasPermissions(String _id, String _subarch) {
		return getPermissions(_id, _subarch) != null;
	}

	public WorkingMemoryPermissions removePermissions(String _id) {
		return m_localPermissions.remove(_id).m_first;
	}
	
	public WorkingMemoryPermissions removePermissions(String _id, String _subarch) {
		if(!m_globalPermissions.containsKey(_subarch)) {
			return null;
		}
		return m_globalPermissions.get(_subarch).remove(_id).m_first;
	}
	
	
	
	
	public boolean needsConsistencyCheck(String _id) {
		return m_localPermissions.get(_id).m_second;
	}
	
	public boolean needsConsistencyCheck(String _id, String _subarch) {
		if(!m_globalPermissions.containsKey(_subarch)) {
			return false;
		}
		return m_globalPermissions.get(_subarch).get(_id).m_second;
	}
	
	public void consistencyChecked(String _id) {
		 m_localPermissions.get(_id).m_second = false;
	}
	
	public void consistencyChecked(String _id, String _subarch) {
		if(!m_globalPermissions.containsKey(_subarch)) {
			return ;
		}
		 m_globalPermissions.get(_subarch).get(_id).m_second = false;
	}
	
	
	
}

File: binding/development/tools/cast/src/c++/cast/core/CASTComponentPermissionsMap.cpp
#include "CASTComponentPermissionsMap.hpp"
#include <CASTUtils.hpp>



namespace cast {
 
  CASTComponentPermissionsMap::CASTComponentPermissionsMap(const std::string & _subarch) {
    //store local permissions in global map for generality
    m_globalPermissions[_subarch] = m_localPermissions;
  }


  void CASTComponentPermissionsMap::setPermissions(const std::string & _id, 
						   const cdl::WorkingMemoryPermissions & _permissions) {
    m_localPermissions[_id].m_first =_permissions;
    m_localPermissions[_id].m_second = true; 	
  }
  
  void CASTComponentPermissionsMap::setPermissions(const std::string & _id, const std::string & _subarch, const cdl::WorkingMemoryPermissions & _permissions) {
      
    m_globalPermissions[_subarch][_id].m_first = _permissions;
    m_globalPermissions[_subarch][_id].m_second = true;
      
  }
    
  const cdl::WorkingMemoryPermissions & CASTComponentPermissionsMap::getPermissions(const std::string & _id) const 
    throw(CASTException) {
    PermissionsMap::const_iterator i = m_localPermissions.find(_id);      
    if(i != m_localPermissions.end()) {
      return i->second.m_first;
    }

    throw(CASTException(exceptionMessage(__HERE__, 
					 "Entry does not have permissions to get: %s", 
					 _id.c_str())));	
    
  }
  
  const cdl::WorkingMemoryPermissions & CASTComponentPermissionsMap::getPermissions(const std::string & _id, 
										    const std::string & _subarch) const 
    throw(CASTException) {

    PermissionsMapMap::const_iterator i = m_globalPermissions.find(_subarch);      
    if(i != m_globalPermissions.end()) {
      PermissionsMap::const_iterator j = i->second.find(_id);
      if(j != i->second.end()) {
	return j->second.m_first;
      }
    }
    throw(CASTException(exceptionMessage(__HERE__, 
					 "Entry does not have permissions to get: %s:%s", 
					 _id.c_str(), _subarch.c_str())));	
  }

  bool CASTComponentPermissionsMap::hasPermissions(const std::string & _id) const {
    PermissionsMap::const_iterator i = m_localPermissions.find(_id);      
    return i != m_localPermissions.end();
  }
    
  bool CASTComponentPermissionsMap::hasPermissions(const std::string & _id, 
						   const std::string & _subarch) const {
    PermissionsMapMap::const_iterator i = m_globalPermissions.find(_subarch);      
    if(i != m_globalPermissions.end()) {
      PermissionsMap::const_iterator j = i->second.find(_id);
      return j != i->second.end();
    }
    else {
      return false;
    }
  }

  void CASTComponentPermissionsMap::removePermissions(const std::string & _id) {
    m_localPermissions.erase(_id);
  }
  
  void CASTComponentPermissionsMap::removePermissions(const std::string & _id, 
						      const std::string & _subarch) {
    if(hasPermissions(_id,_subarch)) {
      m_globalPermissions[_subarch].erase(_id);
    }
  }
  
  bool CASTComponentPermissionsMap::needsConsistencyCheck(const std::string & _id) const {
    if(hasPermissions(_id)) {
      //find should work if above passes
      return m_localPermissions.find(_id)->second.m_second;
    }
    else {
      return false;
    }
  }
    
  bool CASTComponentPermissionsMap::needsConsistencyCheck(const std::string & _id, 
							  const std::string & _subarch) const {
    if(hasPermissions(_id,_subarch)) {
      //find should work if above passes
      return m_globalPermissions.find(_subarch)->second.find(_id)->second.m_second;
    }
    else {
      return false;
    }
  }
  
  void CASTComponentPermissionsMap::consistencyChecked(const std::string & _id) {
    if(hasPermissions(_id)) {
      m_localPermissions[_id].m_second = false;
    }
  }
  
  void CASTComponentPermissionsMap::consistencyChecked(const std::string & _id, 
						       const std::string & _subarch) {     
    if(hasPermissions(_id,_subarch)) {
      m_globalPermissions[_subarch][_id].m_second = false;
    }
  }
  
} 
